devito/data/allocators.py

```diff
@@ -334,6 +337,31 @@ def put_local(self):
         return self._node == 'local'
 
 
+class PetscMemoryAllocator(MemoryAllocator):
+    """
+    """
+    @classmethod
+    def initialize(cls):
+        from devito.petsc.utils import core_metadata
+        metadata = core_metadata()
+        lib_dir = Path(metadata['lib_dirs'][-1])
+
+        try:
+            cls.lib = ctypes.CDLL(lib_dir/'libpetsc.so')
+        except OSError:
+            cls.lib = None
+
+    def _alloc_C_libcall(self, size, ctype):
+        if not self.available():
+            raise RuntimeError("...")
+        c_bytesize = ctypes.c_ulong(size * ctypes.sizeof(ctype))
+        c_pointer = ctypes.cast(ctypes.c_void_p(), ctypes.c_void_p)
+
+        from devito.petsc.memory import op_memory
```

Comment:
There are a couple of imports within functions here, which probably means things need moving around to prevent circular imports